classes/data-stores/ActionScheduler_DBStore.php

```diff
@@ -655,7 +655,7 @@ protected function claim_actions( $claim_id, $limit, \DateTime $before_date = nu
 			$params[] = $group_id;
 		}
 
-		$order    = "ORDER BY attempts ASC, scheduled_date_gmt ASC, action_id ASC LIMIT %d";
+		$order    = "LIMIT %d";
```

Comment:
> Without ORDER BY, the query will still pick the oldest actions (in most cases) because it'll order them based on the action_id which is sequential.

Understood, though, creation date is not 1:1 with scheduled date. 

I'd be just a little worried about cases where honouring that commitment (to process the longest waiting actions relative to their scheduled date/time) is particularly important. There could definitely be lots of cases where it may make no difference, but on the other hand there may be projects and workloads where it is valuable.

Along with the change in table indicies, I wonder if this is something that should exist as an optional change—that is, let site operators opt-in or out of this (and there are various ways we might implement that, ranging from UI-level settings to feature flags or filters). 

Some other (but minor) issues:

- Unless we make a corresponding change to the WP Post Store, this will make the behavior of the two store implementations inconsistent with one another.
- I feel less concerned about dropping ordering on the `attempts` column but it would be interesting to hear from others on this.

Balancing all that, the potential performance improvement is clear. Will flag for further discussion.

Reply:
I see what you mean about ordering, but i wonder if there are many real-world scenarios where this would be an issue? It would only affect stores that rely on actions running in a specific order, but their code creates them in the wrong order. It seems quite unlikely to me.


as a reminder, in our case, the deadlocks was the main issue and it seems the only way to overcome them is by altering this query. With WC/AS being unable to handle those deadlocks, this is the only viable solution and certainly beats any scheduling issues.

Maybe a middle-ground would be to add a filter and a section on docs advising people who get deadlocks to use it? At the moment, we have to hardcode this change to WC and any plugins that ship with AS which is a pain, but having a filter would permanently resolve our deadlocks without affecting other stores who are unaware of them or desperately need them firing in this order?

Reply:
> I see what you mean about ordering, but i wonder if there are many real-world scenarios where this would be an issue? It would only affect stores that rely on actions running in a specific order, but their code creates them in the wrong order. It seems quite unlikely to me.

Yep, my instincts match your own, _but,_ the fact is, we don't know for sure (we don't have perfect visibility into all use cases). Some notes:

- Action Scheduler is not only used alongside WooCommerce—it's also included and used by plugins (and perhaps themes) with no relationship to WooCommerce or even e-commerce in general.
- Re the problem of code creating actions in the wrong order, I'm mostly thinking of a situation where multiple components—be they plugins or themes—which may or may not be related, are independently scheduling actions.

And, again, I'm not presenting this as a blocker—I just want to pause and get thoughts/think of other use cases before moving ahead.

> Maybe a middle-ground would be to add a filter and a section on docs advising people who get deadlocks to use it?

I definitely find that to be an attractive idea.